src/util.h

```diff
@@ -269,6 +269,6 @@ class InsecureRand
     {
         nRz = 36969 * (nRz & 65535) + (nRz >> 16);
         nRw = 18000 * (nRw & 65535) + (nRw >> 16);
-        return ((nRw << 16) + nRz) % nMax;
+        return ((uint32_t)(nRw << 16) + nRz) % nMax;

```

Comment:
I think it should be fixed to:
return (((uint32_t)nRw << 16) + nRz) % nMax;

Reply:
what is difference between (uint32_t)(nRw << 16) and ((uint32_t)nRw << 16)?

Reply:
The best fix should be: return (((int64_t)nRw << 16) + nRz) % nMax;
The difference is:
    -  (uint32_t)(nRw << 16) : Cast after shift
    -  ((uint32_t)nRw << 16): Cast before shift
According to this: https://pvs-studio.com/en/docs/warnings/v629/ 
We must make sure the type of the left argument is 64-bit before the shift, so we must cast it before the sift, that is what I mean.

Reply:
but nRw is also uint32_t type.
so, it is same.

Reply:
nRw is uint32_t type.
But (int64_t)nRw is int64_t type.

Let's do a test:
![image](https://user-images.githubusercontent.com/42252263/120904474-ab75d980-c676-11eb-9457-8f844750ed7f.png)

Reply:
Good catch here Vadym

Reply:
don't use c-style casts, ((static_cast<int64_t>(nRw) << 16) + nRz) % nMax;